README.md
# Frame Streams implementation in Go

https://github.com/farsightsec/golang-framestream

Frame Streams is a lightweight, binary-clean protocol that allows
for the transport of arbitrarily encoded data payload sequences with
minimal framing overhead.

This provides the Golang implementation. The Frame Streams implementation
in C is at https://github.com/farsightsec/fstrm/.

The framestream_dump tool reads a Frame Streams formatted input file
and prints the data frames and frame byte counts.


debian/changelog
golang-framestream (0.3.0-1) debian-fsi; urgency=medium

  * Limit maximum control frame length. Adopt C fstrm limit of 512 bytes,
    not including type+length.
  * Reader: remove log message from library function.
  * Control.go: allow SetContentType(nil) for control frame.
  * Control.go: match content type more efficiently.
  * Add ReadFrame and WriteFrame functions.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Thu, 12 Nov 2020 13:29:10 -0500

golang-framestream (0.2.0-1) debian-fsi; urgency=medium

  * Add I/O timeout for Encoders and Decoders.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Thu, 25 Apr 2019 15:41:46 -0400

golang-framestream (0.1.0-1) debian-fsi; urgency=medium

  * Miscellaneous debian packaging metadata cleanups.

  * Add a README.md.

  * The framestream_dump tool is not installed in this "dev" package.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Thu, 01 Nov 2018 15:11:42 -0400

golang-framestream (0.0.0-1) unstable; urgency=medium

  * Debian packaged release

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Mon, 20 Aug 2018 08:52:42 -0500


pkg_test.go
package framestream

import (
	"bytes"
	"net"
	"testing"
)

var FSContentType = []byte("protobuf:dnstap.Dnstap")

func TestUnidirectional(t *testing.T) {
	b := &bytes.Buffer{}

	enc, err := NewEncoder(b, &EncoderOptions{ContentType: FSContentType})
	if err != nil {
		t.Fatal(err)
		return
	}

	dec, err := NewDecoder(b, &DecoderOptions{ContentType: FSContentType})
	if err != nil {
		t.Fatal(err)
		return
	}

	testEncodeDecode(t, enc, dec)
}
func TestSocket(t *testing.T) {
	var enc *Encoder
	wait := make(chan bool)

	l, err := net.Listen("unix", "sock")
	if err != nil {
		t.Fatal(err)
		return
	}

	go func() {
		server, err := l.Accept()
		if err != nil {
			t.Fatal(err)
			return
		}

		opt := &DecoderOptions{
			ContentType:   FSContentType,
			Bidirectional: true,
		}
		dec, err := NewDecoder(server, opt)
		if err != nil {
			t.Fatal(err)
			return
		}

		testEncodeDecode(t, enc, dec)

		close(wait)
	}()

	if !t.Failed() {
		client, err := net.Dial("unix", "sock")
		if err != nil {
			t.Fatal(err)
			return
		}

		opt := &EncoderOptions{
			ContentType:   FSContentType,
			Bidirectional: true,
		}
		enc, err = NewEncoder(client, opt)
		if err != nil {
			t.Fatal(err)
			return
		}

		<-wait
	}

	if err := l.Close(); err != nil {
		t.Error(err)
	}
}

func testEncodeDecode(t *testing.T, enc *Encoder, dec *Decoder) {
	wants := []string{"frame one", "two", "3"}
	for _, frame := range wants {
		_, err := enc.Write([]byte(frame))
		if err != nil {
			t.Fatalf("encode failed: %s\n", err)
			return
		}
	}
	enc.Flush()
	enc.Close()

	for i, want := range wants {
		frame, err := dec.Decode()
		if err != nil {
			t.Errorf("decode failed: %s\n", err)
		}
		if string(frame) != want {
			t.Errorf("frame %d: wanted: %s, got: %s", i, wants[i], string(frame))
		}
	}
}


framestream_test.go
package framestream_test

import (
	"bytes"
	"fmt"
	"net"
	"sync"
	"testing"
	"time"

	framestream "github.com/farsightsec/golang-framestream"
)

func testDecoder(t *testing.T, dec *framestream.Decoder, nframes int) {
	i := 1
	for {
		tf, err := dec.Decode()
		if err != nil {
			if i < nframes+1 {
				t.Fatalf("testDecoder(%d): %v", i, err)
			}
			if err != framestream.EOF {
				t.Fatalf("unexpected error: %v != EOF", err)
			}
			return
		}
		if i > nframes {
			t.Errorf("extra frame received: %d", i)
		}
		f := make([]byte, i)
		if bytes.Compare(tf, f) != 0 {
			t.Errorf("testDecoder: received %v != %v", tf, f)
		}
		i++
	}
}

func TestUnidirectional(t *testing.T) {
	buf := new(bytes.Buffer)
	enc, err := framestream.NewEncoder(buf, nil)
	if err != nil {
		t.Fatal(err)
	}

	for i := 1; i < 10; i++ {
		b := make([]byte, i)
		if _, err = enc.Write(b); err != nil {
			t.Error(err)
		}
	}
	enc.Close()

	dec, err := framestream.NewDecoder(buf, nil)
	if err != nil {
		t.Fatal(err)
	}
	testDecoder(t, dec, 9)
}

func TestBidirectional(t *testing.T) {
	client, server := net.Pipe()

	go func() {
		dec, err := framestream.NewDecoder(server,
			&framestream.DecoderOptions{
				Bidirectional: true,
			})

		if err != nil {
			t.Fatal(err)
		}
		testDecoder(t, dec, 9)
	}()

	enc, err := framestream.NewEncoder(client,
		&framestream.EncoderOptions{
			Bidirectional: true,
		})
	if err != nil {
		t.Fatal(err)
	}

	for i := 1; i < 10; i++ {
		b := make([]byte, i)
		if _, err := enc.Write(b); err != nil {
			t.Error(err)
		}
	}
	enc.Close()
}

func TestContentTypeMismatch(t *testing.T) {
	buf := new(bytes.Buffer)

	enc, err := framestream.NewEncoder(buf,
		&framestream.EncoderOptions{
			ContentType: []byte("test"),
		})
	if err != nil {
		t.Fatal(err)
	}
	enc.Write([]byte("hello, world"))
	enc.Close()

	_, err = framestream.NewDecoder(buf,
		&framestream.DecoderOptions{
			ContentType: []byte("wrong"),
		})
	if err != framestream.ErrContentTypeMismatch {
		t.Errorf("expected %v, received %v",
			framestream.ErrContentTypeMismatch,
			err)
	}
}

func TestOversizeFrame(t *testing.T) {
	buf := new(bytes.Buffer)
	enc, err := framestream.NewEncoder(buf, nil)
	if err != nil {
		t.Fatal(err)
	}

	enc.Write(make([]byte, 15))
	enc.Close()

	dec, err := framestream.NewDecoder(buf,
		&framestream.DecoderOptions{
			MaxPayloadSize: 10,
		})
	if err != nil {
		t.Fatal(err)
	}
	_, err = dec.Decode()
	if err != framestream.ErrDataFrameTooLarge {
		t.Errorf("data frame too large, received %v", err)
	}
}

func testNew(t *testing.T, bidirectional bool, timeout time.Duration) {
	client, server := net.Pipe()
	wg := &sync.WaitGroup{}
	wg.Add(2)
	done := make(chan bool)

	defer client.Close()
	defer server.Close()

	go func() {
		_, err := framestream.NewDecoder(server,
			&framestream.DecoderOptions{
				Bidirectional: bidirectional,
				Timeout:       timeout,
			})

		if err != nil {
			t.Log("decoder error: ", err)
		}
		wg.Done()
	}()

	go func() {
		_, err := framestream.NewEncoder(client,
			&framestream.EncoderOptions{
				Bidirectional: bidirectional,
				Timeout:       timeout,
			})
		if err != nil {
			t.Log("encoder error: ", err)
		}
		wg.Done()
	}()

	go func() {
		wg.Wait()
		close(done)
	}()

	select {
	case <-done:
	case <-time.After(1 * time.Second):
		t.Fatal("Time's up.")
	}
}

func TestNewBidirectional(t *testing.T) {
	testNew(t, true, 0)
}

func TestNewUnidirectional(t *testing.T) {
	testNew(t, false, 0)
}

func TestNewBidirectionalTimeout(t *testing.T) {
	testNew(t, true, time.Second)
}

func testNegotiate(t *testing.T, rtypes, wtypes [][]byte, expected []byte, ok bool) {
	wc, rc := net.Pipe()
	done := make(chan error)
	defer wc.Close()
	defer rc.Close()
	go func() {
		w, err := framestream.NewWriter(wc, &framestream.WriterOptions{
			Bidirectional: true,
			ContentTypes:  wtypes,
		})
		defer close(done)
		if err != nil {
			done <- err
			wc.Close()
			return
		}
		if bytes.Compare(expected, w.ContentType()) != 0 {
			done <- fmt.Errorf("writer content type %s != %s",
				w.ContentType(), expected)
		}
	}()

	r, err := framestream.NewReader(rc, &framestream.ReaderOptions{
		Bidirectional: true,
		ContentTypes:  rtypes,
	})
	if err != nil {
		if ok {
			t.Errorf("NewReader error: %v", err)
		}
		rc.Close()
	} else {
		if !ok {
			t.Errorf("NewReader did not fail as expected")
		}
		if bytes.Compare(expected, r.ContentType()) != 0 {
			t.Errorf("reader content type %s != %s",
				r.ContentType(), expected)
		}
	}
	err = <-done
	if ok && err != nil {
		t.Errorf("Writer error: %v", err)
	}
	if !ok && err == nil {
		t.Errorf("Writer did not fail as expected")
	}
}

func contentTypes(types ...string) [][]byte {
	b := make([][]byte, 0)
	for _, t := range types {
		b = append(b, []byte(t))
	}
	return b
}

func TestContentTypes(t *testing.T) {
	t.Log("reader: nil, writer: nil, expect: nil")
	testNegotiate(t, nil, nil, nil, true)
	t.Log("reader: type1, writer: type1, expect: type1")
	testNegotiate(t,
		contentTypes("type1"),
		contentTypes("type1"),
		[]byte("type1"), true)
	t.Log("reader: type1, writer: type2, expect: error")
	testNegotiate(t,
		contentTypes("type1"),
		contentTypes("type2"),
		nil, false)
	t.Log("reader: type1, type2, type3, writer: type4, type3, type2, expect: type2")
	testNegotiate(t,
		contentTypes("type1", "type2", "type3"),
		contentTypes("type4", "type3", "type2"),
		[]byte("type2"), true)
}


Reader.go
/*
 * Copyright (c) 2014 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package framestream

import (
	"bufio"
	"encoding/binary"
	"io"
	"io/ioutil"
	"log"
	"time"
)

type ReaderOptions struct {
	// The ContentTypes accepted by the Reader. May be left unset for no
	// content negotiation. If the corresponding Writer offers a disjoint
	// set of ContentTypes, NewReader() will return ErrContentTypeMismatch.
	ContentTypes [][]byte
	// If Bidirectional is true, the underlying io.Reader must be an
	// io.ReadWriter, and the Reader will engage in a bidirectional
	// handshake with its peer to establish content type and communicate
	// shutdown.
	Bidirectional bool
	// Timeout gives the timeout for reading the initial handshake messages
	// from the peer and writing response messages if Bidirectional. It is
	// only effective for underlying Readers satisfying net.Conn.
	Timeout time.Duration
}

// Reader reads data frames from an underlying io.Reader using the Frame
// Streams framing protocol.
type Reader struct {
	contentType   []byte
	bidirectional bool
	r             *bufio.Reader
	w             *bufio.Writer
	stopped       bool
}

// NewReader creates a Frame Streams Reader reading from the given io.Reader
// with the given ReaderOptions.
func NewReader(r io.Reader, opt *ReaderOptions) (*Reader, error) {
	if opt == nil {
		opt = &ReaderOptions{}
	}
	tr := timeoutReader(r, opt)
	reader := &Reader{
		bidirectional: opt.Bidirectional,
		r:             bufio.NewReader(tr),
		w:             nil,
	}

	if len(opt.ContentTypes) > 0 {
		reader.contentType = opt.ContentTypes[0]
	}

	var cf ControlFrame
	if opt.Bidirectional {
		w, ok := tr.(io.Writer)
		if !ok {
			return nil, ErrType
		}
		reader.w = bufio.NewWriter(w)

		// Read the ready control frame.
		err := cf.DecodeTypeEscape(reader.r, CONTROL_READY)
		if err != nil {
			log.Println("DecodeTypeEscape: ", err)
			return nil, err
		}

		// Check content type.
		if t, ok := cf.ChooseContentType(opt.ContentTypes); ok {
			reader.contentType = t
		} else {
			return nil, ErrContentTypeMismatch
		}

		// Send the accept control frame.
		accept := ControlAccept
		accept.SetContentType(reader.contentType)
		err = accept.EncodeFlush(reader.w)
		if err != nil {
			return nil, err
		}
	}

	// Read the start control frame.
	err := cf.DecodeTypeEscape(reader.r, CONTROL_START)
	if err != nil {
		return nil, err
	}

	// Disable the read timeout to prevent killing idle connections.
	disableReadTimeout(tr)

	// Check content type.
	if !cf.MatchContentType(reader.contentType) {
		return nil, ErrContentTypeMismatch
	}

	return reader, nil
}

// ReadFrame reads a data frame into the supplied buffer, returning its length.
// If the frame is longer than the supplied buffer, Read returns
// ErrDataFrameTooLarge and discards the frame. Subsequent calls to Read()
// after this error may succeed.
func (r *Reader) ReadFrame(b []byte) (length int, err error) {
	if r.stopped {
		return 0, EOF
	}

	for length == 0 {
		length, err = r.readFrame(b)
		if err != nil {
			return
		}
	}

	return
}

// ContentType returns the content type negotiated with the Writer.
func (r *Reader) ContentType() []byte {
	return r.contentType
}

func (r *Reader) readFrame(b []byte) (int, error) {
	// Read the frame length.
	var frameLen uint32
	err := binary.Read(r.r, binary.BigEndian, &frameLen)
	if err != nil {
		return 0, err
	}

	if frameLen > uint32(len(b)) {
		io.CopyN(ioutil.Discard, r.r, int64(frameLen))
		return 0, ErrDataFrameTooLarge
	}

	if frameLen == 0 {
		// This is a control frame.
		var cf ControlFrame
		err = cf.Decode(r.r)
		if err != nil {
			return 0, err
		}
		if cf.ControlType == CONTROL_STOP {
			r.stopped = true
			if r.bidirectional {
				ff := &ControlFrame{ControlType: CONTROL_FINISH}
				err = ff.EncodeFlush(r.w)
				if err != nil {
					return 0, err
				}
			}
			return 0, EOF
		}
		return 0, err
	}

	return io.ReadFull(r.r, b[0:frameLen])
}


Writer.go
/*
 * Copyright (c) 2014 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package framestream

import (
	"bufio"
	"encoding/binary"
	"io"
	"time"
)

type WriterOptions struct {
	// The ContentTypes available to be written to the Writer. May be
	// left unset for no content negotiation. If the Reader requests a
	// disjoint set of content types, NewEncoder() will return
	// ErrContentTypeMismatch.
	ContentTypes [][]byte
	// If Bidirectional is true, the underlying io.Writer must be an
	// io.ReadWriter, and the Writer will engage in a bidirectional
	// handshake with its peer to establish content type and communicate
	// shutdown.
	Bidirectional bool
	// Timeout gives the timeout for writing both control and data frames,
	// and for reading responses to control frames sent. It is only
	//  effective for underlying Writers satisfying net.Conn.
	Timeout time.Duration
}

// A Writer writes data frames to a Frame Streams file or connection.
type Writer struct {
	contentType []byte
	w           *bufio.Writer
	r           *bufio.Reader
	opt         WriterOptions
	buf         []byte
}

// NewWriter returns a Frame Streams Writer using the given io.Writer and options.
func NewWriter(w io.Writer, opt *WriterOptions) (writer *Writer, err error) {
	if opt == nil {
		opt = &WriterOptions{}
	}
	w = timeoutWriter(w, opt)
	writer = &Writer{
		w:   bufio.NewWriter(w),
		opt: *opt,
	}

	if opt.ContentTypes != nil {
		writer.contentType = opt.ContentTypes[0]
	}

	if opt.Bidirectional {
		r, ok := w.(io.Reader)
		if !ok {
			return nil, ErrType
		}
		writer.r = bufio.NewReader(r)
		ready := ControlReady
		ready.SetContentTypes(opt.ContentTypes)
		if err = ready.EncodeFlush(writer.w); err != nil {
			return
		}

		var accept ControlFrame
		if err = accept.DecodeTypeEscape(writer.r, CONTROL_ACCEPT); err != nil {
			return
		}

		if t, ok := accept.ChooseContentType(opt.ContentTypes); ok {
			writer.contentType = t
		} else {
			return nil, ErrContentTypeMismatch
		}
	}

	// Write the start control frame.
	start := ControlStart
	start.SetContentType(writer.contentType)
	err = start.EncodeFlush(writer.w)
	if err != nil {
		return
	}

	return
}

// ContentType returns the content type negotiated with Reader.
func (w *Writer) ContentType() []byte {
	return w.contentType
}

// Close shuts down the Frame Streams stream by writing a CONTROL_STOP message.
// If the Writer is Bidirectional, Close will wait for an acknowledgement
// (CONTROL_FINISH) from its peer.
func (w *Writer) Close() (err error) {
	err = ControlStop.EncodeFlush(w.w)
	if err != nil || !w.opt.Bidirectional {
		return
	}

	var finish ControlFrame
	return finish.DecodeTypeEscape(w.r, CONTROL_FINISH)
}

// WriteFrame writes the given frame to the underlying io.Writer with Frame Streams
// framing.
func (w *Writer) WriteFrame(frame []byte) (n int, err error) {
	err = binary.Write(w.w, binary.BigEndian, uint32(len(frame)))
	if err != nil {
		return
	}
	return w.w.Write(frame)
}

// Flush ensures that any buffered data frames are written to the underlying
// io.Writer.
func (w *Writer) Flush() error {
	return w.w.Flush()
}


Control.go
package framestream

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"io"
)

const CONTROL_ACCEPT = 0x01
const CONTROL_START = 0x02
const CONTROL_STOP = 0x03
const CONTROL_READY = 0x04
const CONTROL_FINISH = 0x05

const CONTROL_FIELD_CONTENT_TYPE = 0x01

type ControlFrame struct {
	ControlType  uint32
	ContentTypes [][]byte
}

var ControlStart = ControlFrame{ControlType: CONTROL_START}
var ControlStop = ControlFrame{ControlType: CONTROL_STOP}
var ControlReady = ControlFrame{ControlType: CONTROL_READY}
var ControlAccept = ControlFrame{ControlType: CONTROL_ACCEPT}
var ControlFinish = ControlFrame{ControlType: CONTROL_FINISH}

func (c *ControlFrame) Encode(w io.Writer) (err error) {
	var buf bytes.Buffer
	err = binary.Write(&buf, binary.BigEndian, c.ControlType)
	if err != nil {
		return
	}
	for _, ctype := range c.ContentTypes {
		err = binary.Write(&buf, binary.BigEndian, uint32(CONTROL_FIELD_CONTENT_TYPE))
		if err != nil {
			return
		}

		err = binary.Write(&buf, binary.BigEndian, uint32(len(ctype)))
		if err != nil {
			return
		}

		_, err = buf.Write(ctype)
		if err != nil {
			return
		}
	}

	err = binary.Write(w, binary.BigEndian, uint32(0))
	if err != nil {
		return
	}
	err = binary.Write(w, binary.BigEndian, uint32(buf.Len()))
	if err != nil {
		return
	}
	_, err = buf.WriteTo(w)
	return
}

func (c *ControlFrame) EncodeFlush(w *bufio.Writer) error {
	if err := c.Encode(w); err != nil {
		return err
	}
	return w.Flush()
}

func (c *ControlFrame) Decode(r io.Reader) (err error) {
	var cflen uint32
	err = binary.Read(r, binary.BigEndian, &cflen)
	if err != nil {
		return
	}

	err = binary.Read(r, binary.BigEndian, &c.ControlType)
	if err != nil {
		return
	}

	cflen -= 4
	if cflen > 0 {
		cfields := make([]byte, int(cflen))
		_, err = io.ReadFull(r, cfields)
		if err != nil {
			return
		}

		for len(cfields) > 8 {
			cftype := binary.BigEndian.Uint32(cfields[:4])
			cfields = cfields[4:]
			if cftype != CONTROL_FIELD_CONTENT_TYPE {
				return ErrDecode
			}

			cflen := int(binary.BigEndian.Uint32(cfields[:4]))
			cfields = cfields[4:]
			if cflen > len(cfields) {
				return ErrDecode
			}

			c.ContentTypes = append(c.ContentTypes, cfields[:cflen])
			cfields = cfields[cflen:]
		}

		if len(cfields) > 0 {
			return ErrDecode
		}
	}
	return
}

func (c *ControlFrame) DecodeEscape(r io.Reader) error {
	var zero uint32
	err := binary.Read(r, binary.BigEndian, &zero)
	if err != nil {
		return err
	}
	if zero != 0 {
		return ErrDecode
	}
	return c.Decode(r)
}

func (c *ControlFrame) DecodeTypeEscape(r io.Reader, ctype uint32) error {
	err := c.DecodeEscape(r)
	if err != nil {
		return err
	}

	if ctype != c.ControlType {
		return ErrDecode
	}

	return nil
}

// ChooseContentType selects a content type from the ControlFrame which
// also exists in the supplied ctypes. Preference is given to values occurring
// earliest in ctypes.
//
// ChooseContentType returns the chosen content type, which may be nil, and
// a bool value indicating whether a matching type was found.
//
// If either the ControlFrame types or ctypes is empty, ChooseContentType
// returns nil as a matching content type.
func (c *ControlFrame) ChooseContentType(ctypes [][]byte) (typ []byte, found bool) {
	if c.ContentTypes == nil || ctypes == nil {
		return nil, true
	}
	tm := make(map[string]bool)
	for _, cfctype := range c.ContentTypes {
		tm[string(cfctype)] = true
	}
	for _, ctype := range ctypes {
		if tm[string(ctype)] {
			return ctype, true
		}
	}
	return nil, false
}

func (c *ControlFrame) MatchContentType(ctype []byte) bool {
	for _, cfctype := range c.ContentTypes {
		if bytes.Compare(ctype, cfctype) == 0 {
			return true
		}
	}
	return len(c.ContentTypes) == 0
}

func (c *ControlFrame) SetContentTypes(ctypes [][]byte) {
	c.ContentTypes = ctypes
}

func (c *ControlFrame) SetContentType(ctype []byte) {
	if ctype != nil {
		c.SetContentTypes([][]byte{ctype})
	} else {
		c.ContentTypes = nil
	}
}


Decoder.go
/*
 * Copyright (c) 2014 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package framestream

import (
	"io"
	"time"
)

// DecoderOptions specifies configuration for a framestream Decoder.
type DecoderOptions struct {
	// MaxPayloadSize is the largest frame size accepted by the Decoder.
	//
	// If the Frame Streams Writer sends a frame in excess of this size,
	// Decode() will return the error ErrDataFrameTooLarge. The Decoder
	// attempts to recover from this error, so calls to Decode() after
	// receiving this error may succeed.
	MaxPayloadSize uint32
	// The ContentType expected by the Decoder. May be left unset for no
	// content negotiation. If the Writer requests a different content type,
	// NewDecoder() will return ErrContentTypeMismatch.
	ContentType []byte
	// If Bidirectional is true, the underlying io.Reader must be an
	// io.ReadWriter, and the Decoder will engage in a bidirectional
	// handshake with its peer to establish content type and communicate
	// shutdown.
	Bidirectional bool
	// Timeout gives the timeout for reading the initial handshake messages
	// from the peer and writing response messages if Bidirectional. It is
	// only effective for underlying Readers satisfying net.Conn.
	Timeout time.Duration
}

// A Decoder decodes Frame Streams frames read from an underlying io.Reader.
//
// It is provided for compatibility. Use Reader instead.
type Decoder struct {
	buf []byte
	r   *Reader
}

// NewDecoder returns a Decoder using the given io.Reader and options.
func NewDecoder(r io.Reader, opt *DecoderOptions) (*Decoder, error) {
	if opt == nil {
		opt = &DecoderOptions{}
	}
	if opt.MaxPayloadSize == 0 {
		opt.MaxPayloadSize = DEFAULT_MAX_PAYLOAD_SIZE
	}
	ropt := &ReaderOptions{
		Bidirectional: opt.Bidirectional,
		Timeout:       opt.Timeout,
	}
	if opt.ContentType != nil {
		ropt.ContentTypes = append(ropt.ContentTypes, opt.ContentType)
	}
	dr, err := NewReader(r, ropt)
	if err != nil {
		return nil, err
	}
	dec := &Decoder{
		buf: make([]byte, opt.MaxPayloadSize),
		r:   dr,
	}
	return dec, nil
}

// Decode returns the data from a Frame Streams data frame. The slice returned
// is valid until the next call to Decode.
func (dec *Decoder) Decode() (frameData []byte, err error) {
	n, err := dec.r.Read(dec.buf)
	if err != nil {
		return nil, err
	}
	return dec.buf[:n], nil
}


Encoder.go
/*
 * Copyright (c) 2014 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package framestream

import (
	"io"
	"time"
)

// EncoderOptions specifies configuration for an Encoder
type EncoderOptions struct {
	// The ContentType of the data sent by the Encoder. May be left unset
	// for no content negotiation. If the Reader requests a different
	// content type, NewEncoder() will return ErrContentTypeMismatch.
	ContentType []byte
	// If Bidirectional is true, the underlying io.Writer must be an
	// io.ReadWriter, and the Encoder will engage in a bidirectional
	// handshake with its peer to establish content type and communicate
	// shutdown.
	Bidirectional bool
	// Timeout gives the timeout for writing both control and data frames,
	// and for reading responses to control frames sent. It is only
	//  effective for underlying Writers satisfying net.Conn.
	Timeout time.Duration
}

// An Encoder sends data frames over a FrameStream Writer.
//
// Encoder is provided for compatibility, use Writer instead.
type Encoder struct {
	*Writer
}

// NewEncoder creates an Encoder writing to the given io.Writer with the given
// EncoderOptions.
func NewEncoder(w io.Writer, opt *EncoderOptions) (enc *Encoder, err error) {
	if opt == nil {
		opt = &EncoderOptions{}
	}
	wopt := &WriterOptions{
		Bidirectional: opt.Bidirectional,
		Timeout:       opt.Timeout,
	}
	if opt.ContentType != nil {
		wopt.ContentTypes = append(wopt.ContentTypes, opt.ContentType)
	}
	writer, err := NewWriter(w, wopt)
	if err != nil {
		return nil, err
	}
	return &Encoder{Writer: writer}, nil
}


decoder.go
/*
 * Copyright (c) 2014 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package framestream

import "bufio"
import "bytes"
import "encoding/binary"
import "errors"
import "io"

const CONTROL_ACCEPT                = 0x01
const CONTROL_START                 = 0x02
const CONTROL_STOP                  = 0x03

const CONTROL_FIELD_CONTENT_TYPE    = 0x01

const DEFAULT_MAX_PAYLOAD_SIZE      = 1048576
const MAX_CONTROL_FRAME_SIZE        = 512

type DecoderOptions struct {
    MaxPayloadSize  uint32
    ContentType     []byte
}

type Decoder struct {
    ControlStart    []byte
    ControlStop     []byte
    ContentType     []byte
    buf             []byte
    opt             DecoderOptions
    rd              *bufio.Reader
    stopped         bool
}

var ErrContentTypeMismatch = errors.New("content type mismatch")
var ErrDataFrameTooLarge = errors.New("data frame too large")
var ErrShortRead = errors.New("short read")
var ErrDecode = errors.New("decoding error")

func NewDecoder(rd io.Reader, opt *DecoderOptions) (dec *Decoder, err error) {
    if opt == nil {
        opt = &DecoderOptions{}
    }
    if opt.MaxPayloadSize == 0 {
        opt.MaxPayloadSize = DEFAULT_MAX_PAYLOAD_SIZE
    }
    dec = &Decoder{
        buf:            make([]byte, opt.MaxPayloadSize),
        opt:            *opt,
        rd:             bufio.NewReader(rd),
    }

    // Read the start control frame.
    err = dec.readEscape()
    if err != nil {
        return
    }
    err = dec.readControlStart()
    if err != nil {
        return
    }

    // Check content type.
    if dec.ContentType != nil && dec.opt.ContentType != nil {
        if bytes.Compare(dec.ContentType, dec.opt.ContentType) != 0 {
            err = ErrContentTypeMismatch
            return
        }
    }

    return
}

func (dec *Decoder) readBE32() (val uint32, err error) {
    err = binary.Read(dec.rd, binary.BigEndian, &val)
    if err != nil {
        return 0, err
    }
    return
}

func (dec *Decoder) readEscape() (err error) {
    escape, err := dec.readBE32()
    if err != nil || escape != 0 {
        dec.stopped = true
        return
    }
    if escape != 0 {
        err = ErrDecode
        return
    }
    return
}

func (dec *Decoder) readControlFrame() (controlFrameData []byte,
                                        controlFrameType uint32,
                                        err error) {
    // Read the control frame length.
    controlFrameLen, err := dec.readBE32()
    if err != nil {
        return
    }

    // Enforce limits on control frame size.
    if controlFrameLen < 4 || controlFrameLen > MAX_CONTROL_FRAME_SIZE {
        err = ErrDecode
        return
    }

    // Read the control frame.
    controlFrameData = make([]byte, controlFrameLen)
    n, err := io.ReadFull(dec.rd, controlFrameData)
    if err != nil || uint32(n) != controlFrameLen {
        return
    }

    // Read the control frame type.
    p := bytes.NewBuffer(controlFrameData[0:4])
    err = binary.Read(p, binary.BigEndian, &controlFrameType)
    if err != nil {
        return
    }

    return
}

func (dec *Decoder) parseControlStart() {
    // Require that the control frame is long enough for three
    // 32-bit BE integers.
    if len(dec.ControlStart) < 12 {
        return
    }

    // Control type: 32-bit BE integer.
    controlFrameType := binary.BigEndian.Uint32(dec.ControlStart[0:4])
    if controlFrameType != CONTROL_START {
        return
    }

    // FSTRM_CONTROL_FIELD_CONTENT_TYPE: 32-bit BE integer.
    controlFieldType := binary.BigEndian.Uint32(dec.ControlStart[4:8])
    if controlFieldType != CONTROL_FIELD_CONTENT_TYPE {
        return
    }

    // Length of content type string: 32-bit BE integer.
    lenContentType := binary.BigEndian.Uint32(dec.ControlStart[8:12])
    if lenContentType > MAX_CONTROL_FRAME_SIZE {
        return
    }

    // The content type string itself: 'lenContentType' bytes.
    dec.ContentType = make([]byte, lenContentType)
    copy(dec.ContentType, dec.ControlStart[12:12+lenContentType])
}

func (dec *Decoder) readControlStart() (err error) {
    controlFrameData, controlFrameType, err := dec.readControlFrame()
    if err != nil {
        return
    }
    if controlFrameType != CONTROL_START {
        return ErrDecode
    }
    dec.ControlStart = controlFrameData
    dec.parseControlStart()
    return
}

func (dec *Decoder) readFrame(frameLen uint32) (err error) {
    // Enforce limits on frame size.
    if frameLen > dec.opt.MaxPayloadSize {
        err = ErrDataFrameTooLarge
        return
    }

    // Read the frame.
    n, err := io.ReadFull(dec.rd, dec.buf[0:frameLen])
    if err != nil || uint32(n) != frameLen {
        return
    }
    return
}

func (dec *Decoder) Decode() (frameData []byte, err error) {
    if dec.stopped {
        err = io.EOF
        return
    }

    // Read the frame length.
    frameLen, err := dec.readBE32()
    if err != nil {
        return
    }

    if frameLen == 0 {
        // This is a control frame.
        controlFrameData, controlFrameType, err := dec.readControlFrame()
        if controlFrameType == CONTROL_STOP {
            dec.ControlStop = controlFrameData
            dec.stopped = true
            return nil, io.EOF
        }
        if err != nil {
            return nil, err
        }

    } else {
        // This is a data frame.
        err = dec.readFrame(frameLen)
        if err != nil {
            return
        }
        frameData = dec.buf[0:frameLen]
    }

    return
}
